pkgs/desktops/lomiri/services/lomiri-url-dispatcher/default.nix

```diff
@@ -0,0 +1,161 @@
+, lomiri-ui-toolkit
+, makeWrapper
+, pkg-config
+, python3
+, qtbase
+, qtdeclarative
+, qtwayland
+, runtimeShell
+, sqlite
+, systemd
+, wrapQtAppsHook
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "lomiri-url-dispatcher";
+  version = "0.1.3";
+
+  src = fetchFromGitLab {
+    owner = "ubports";
+    repo = "development/core/lomiri-url-dispatcher";
+    rev = finalAttrs.version;
+    hash = "sha256-kde/HzhBHxTeyc2TCUJwpG7IfC8doDd/jNMF8KLM7KU=";
+  };
+
+  outputs = [
+    "out"
+    "dev"
+  ];
+
+  patches = [
+    # Fix case-sensitivity in tests
+    # Remove when https://gitlab.com/ubports/development/core/lomiri-url-dispatcher/-/merge_requests/8 merged & in release
+    (fetchpatch {
+      url = "https://gitlab.com/sunweaver/lomiri-url-dispatcher/-/commit/ebdd31b9640ca243e90bc7b8aca7951085998bd8.patch";
+      hash = "sha256-g4EohB3oDcWK4x62/3r/g6CFxqb7/rdK51+E/Fji1Do=";
+    })
+  ];
+
+  postPatch = ''
+    substituteInPlace data/CMakeLists.txt \
+      --replace "\''${SYSTEMD_USER_UNIT_DIR}" "\''${CMAKE_INSTALL_LIBDIR}/systemd/user"
+
+    substituteInPlace tests/url_dispatcher_testability/CMakeLists.txt \
+      --replace "\''${PYTHON_PACKAGE_DIR}" "$out/${python3.sitePackages}"
+
+    # Update URI handler database whenever new url-handler is installed system-wide
+    substituteInPlace data/lomiri-url-dispatcher-update-system-dir.*.in \
+      --replace '@CMAKE_INSTALL_FULL_DATAROOTDIR@' '/run/current-system/sw/share'
+  '' + lib.optionalString finalAttrs.finalPackage.doCheck ''
+    patchShebangs tests/test-sql.sh
+  '';
+
+  strictDeps = true;
+
+  nativeBuildInputs = [
+    cmake
+    glib # for gdbus-codegen
+    intltool
+    makeWrapper
+    pkg-config
+    (python3.withPackages (ps: with ps; [
+      setuptools
+    ] ++ lib.optionals finalAttrs.finalPackage.doCheck [
+      python-dbusmock
+    ]))
+    wrapQtAppsHook
+  ];
+
+  buildInputs = [
+    cmake-extras
+    dbus-test-runner
+    glib
+    gtest
+    json-glib
+    libapparmor
+    lomiri-app-launch
+    lomiri-ui-toolkit
+    qtdeclarative
+    sqlite
+    systemd
+    libxkbcommon
+  ];
+
+  nativeCheckInputs = [
+    dbus
+    sqlite
+  ];
+
+  cmakeFlags = [
+    "-DLOCAL_INSTALL=ON"
+    "-Denable_mirclient=OFF"
+  ];
+
+  doCheck = stdenv.buildPlatform.canExecute stdenv.hostPlatform;
+
+  # Tests work with an sqlite db, cannot handle >1 test at the same time
+  enableParallelChecking = false;
+
+  dontWrapQtApps = true;
+
+  preFixup = ''
+    wrapProgram $out/bin/lomiri-url-dispatcher-dump \
+      --prefix PATH : ${lib.makeBinPath [ sqlite ]}
+
+    # Move from qmlscene call in desktop file to easier-to-wrap script
+    guiScript=$out/bin/lomiri-url-dispatcher-gui
+    guiExec=$(grep 'Exec=' $out/share/applications/lomiri-url-dispatcher-gui.desktop | cut -d'=' -f2-)
+
+    cat <<EOF >$guiScript
+    #!${runtimeShell}
+    $guiExec
+    EOF
+    chmod +x $guiScript
+    substituteInPlace $out/share/applications/lomiri-url-dispatcher-gui.desktop \
```

Comment:
It seems `Icon` is a store path. Some desktop environments like Gnome copy desktop entry files to a startup applications directory, which is a problem when the referenced store paths get garbage collected.

Please move/copy/link the icon `share/pixmaps` and refer to it by name.

Reply:
> Some desktop environments like Gnome copy desktop entry files to a startup applications directory, which is a problem when the referenced store paths get garbage collected.

Shouldn't those paths only get GC'd when the packages aren't considered to be "installed" anymore? Or are they copied over once, and never again?

> Please move/copy/link the icon `share/pixmaps` and refer to it by name.

The icon is a vector graphic, not a pixel-based image format. Should it rather go to `share/icons/hicolor/scalable/apps` then?

As a soon-to-be-relevant follow-up to all of this: Many of the graphical Lomiri applications will have paths to extra graphics defined in their desktop files, i.e. for splash screens. Where should we move/link those graphics to avoid this issue, `scalable/apps` as well?

Reply:
> Or are they copied over once, and never again?

To my understanding, at the time they are selected ot run on startup in e.g. Gnome Tweaks

> Should it rather go to share/icons/hicolor/scalable/apps then?

:+1:

> Where should we move/link those graphics?

Going by [this search](https://github.com/search?type=code&q=X-Lomiri-Splash-Image%3D) it seems possible to make the splash relative, but to _what_ I do not know. Perhaps 
Critically `Exec` and `Icon` should not refer to the store, but whether `X-Lomiri-Splash-Image` is affected the same way as `Icon` i can't say. If the startup applications never display a splash, then it might be okay. An option is to await and see if it becomes a problem.